src/audio/pipeline.c

```diff
@@ -145,26 +145,56 @@ struct pipeline_walk_context {
 	/**< pipelines to be scheduled after trigger walk */
 	struct list_item pipelines;
 	/*
-	 * If this flag is set, pipeline_for_each_comp() will skip all
+	 * If this flag is set, pipeline_for_each_element() will skip all
 	 * incompletely initialised components, i.e. those, whose .pipeline ==
 	 * NULL. Such coomponents should not be skipped during initialisation
 	 * and clean up, but they should be skipped during streaming.
 	 */
 	bool skip_incomplete;
+
+	/**
+	 * when true, then comp_func is called after processing component
+	 * buffers and following components, otherwise comp_func is called at
+	 * first place.
+	 *
+	 * Example:
+	 * For topology: c1 -> b2 -> c3 -> b4 -> c5
+	 * context functions will be applied in order:
+	 *   false: c1, b2, c3, b4, c5
+	 *    true: c5, b4, c3, b2, c1
+	 */
+	bool call_func_on_exit;
 };
 
-/* Generic method for walking the graph upstream or downstream.
+/**
+ * Generic method for walking the graph upstream or downstream.
  * It requires function pointer for recursion.
+ * @param current initial node of components graph
+ * @param calling_buf trace for ancestor, passed as one of ctx.comp_func() argument
+ * @param ctx target function pointers holder and graph walk-style description
+ * @param dir PPL_DIR_DOWNSTREAM when walking through sinks, PPL_DIR_UPSTREAM otherwise
```

Comment:
ok, see my above comment. Now I'm confused since it looks like we have two ways to set calling direction ?